File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> ans;
        unordered_map<int,int> mp;

        // return the k most frequent elements

        // Method 1 - create map , put it inot priority queue and get the top k
        // most frequent elements

        // Method 2 - // nums[i]

        // create map

        int n = nums.size();

        for (auto x : nums) {
            mp[x]++;
        }

        /*
            1 - 3
            2 - 2
            3 - 1
        */

        // put it into bucket sort

        vector<vector<int>> arr(n+1);

        for (auto& [key, val] : mp) {
            arr[val].push_back(key);
        }

        for (int i = n ; i >= 0; i--) {
            if (arr[i].size() != 0) {
                if (k > 0) {
                    for (auto x : arr[i]) {
                        if (k > 0) {
                            ans.push_back(x);
                            k--;
                        }
                    }
                }
            }
        }

        return ans;
    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # find the top k frequent elements 

        my_dict = {}
        n = len(nums)

        ans = []

        for i in range(n):
            my_dict[nums[i]]= my_dict.get(nums[i],0)+1
        

        # now have heap 

        # import heapq

        # heap = []

        # # max heap 

        # for key,value in my_dict.items():
        #     heapq.heappush(heap,[-value,key])
        

        # while k:
        #     [value,key] = heapq.heappop(heap)
        #     ans.append(key)
        #     k-=1


        ## Without heap solution 

        # using bucket sort 

        ls = [[] for x in range(n)]

        for key,value in my_dict.items():
            ls[value-1].append(key)
        

        print(ls)

        for i in range(n-1,-1,-1): # iterating in reverse 
            if len(ls[i])>0:
                for j in range (len(ls[i])):
                    if(k>0):
                        ans.append(ls[i][j])
                        k-=1
                    elif(k<=0):
                        return ans

                



        

        return ans
